### ISSUE TITLE
[BUG] Unable to join a cluster

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#1993]
**Describe the bug**

This was observed after https://github.com/harvester/harvester-installer/pull/229.
Harvester installer vagrant tests might fail because the second node can't join the cluster.
Example: https://ci.harvesterhci.io/job/harvester-vagrant-installation-test/279/console

I also repro this once in my local virtual machine envrionment.

**To Reproduce**
1. Create a cluster
2. Join a second node, the node never appears in `kubectl get nodes`.

**Expected behavior**

New nodes should join.

**Support bundle**
[supportbundle_1f3d106d-8b3c-409a-a2af-700cc25058b4_2022-03-10T08-39-52Z.zip](https://github.com/harvester/harvester/files/8221768/supportbundle_1f3d106d-8b3c-409a-a2af-700cc25058b4_2022-03-10T08-39-52Z.zip)

**Environment:**
 - Harvester ISO version: master ISO (2022 Mar. 9th).

**Additional context**
The node can't join because the cluster is still under provisioning.

rancher log:

```
rancher-7bf75cf475-56bnm 2022/03/10 05:37:29 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
rancher-7bf75cf475-56bnm 2022/03/10 05:37:30 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
rancher-7bf75cf475-56bnm 2022/03/10 05:37:31 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
rancher-7bf75cf475-56bnm 2022/03/10 05:37:31 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
rancher-7bf75cf475-56bnm 2022/03/10 05:37:31 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
rancher-7bf75cf475-56bnm 2022/03/10 05:37:31 [INFO] [planner] rkecluster fleet-local/local: waiting: waiting for join url to be available on bootstrap node
```

By checking rancher's code, it might be that the `rke.cattle.io/join-url` annotation is empty in machine plans:

```
# kubectl get secrets -n fleet-local  | grep rke.cattle.io/machine-plan  | awk '{print $1}'
custom-5402c6b1ed11-machine-plan
custom-bd3a88252173-machine-plan
```

```
harvester-node-0:/var/log # kubectl get secrets -n fleet-local  | grep rke.cattle.io/machine-plan  | awk '{print $1}' | xargs -I {} kubectl get secret/{} -o jsonpath='{.metadata.annotations}' -n fleet-local
{"objectset.rio.cattle.io/applied":"H4sIAAAAAAAA/4xSQY+bPBD9K5/mjPPFBAgg9VK1pz1UWlW9VD0M9pAAxkb2ENpd5b9XsInUJJu0R4b33jy/ea/QE6NGRihfAa11jNw4G+ZPV7WkOBCvfONWCpkNrRr3f6OhBN+R6FHtG0sQ3YW6yZIXu0P3xvjjz0FG/z01Vn94fvr80TkO7HH4q5LFnqAENQZ2vUiTdayySpKW8p+oYUA182tDxMI4hQaiK2Ota6wYvYES4BiB8rQk8rXpKTD2A5R2NCYCgxWZJSdlxsDkVz9Fl4dZ4jQ4uz3vuedvj2EPJcS0liqN46TSGcUSldpgnqw3mOUyxwK3scyLDSY3ju+suwI5y94ZMRi0JLwzM5T9SDdIYqUfAk5nF0sP+pTbHbPK8zqeXnZ1VfSHzraVb/WLqrZqV2+3+xeTHdqB0y7pp3aa0vSu5sMDX1Im5zvyF06PEdwXONd1iQDekO8XYinMM9XkySoKUH5/BRyab+RD4+w7VYYIKuNU92UmfiJDvOBmTxGckjfkz5OusXN0V81/+PRxCVuncb6NMy2KPMlFIvVGFLXUAot8XcRaZlKt4fjjGAH/GujG6EUAx98BAAD//1RTtk78AwAA","objectset.rio.cattle.io/id":"rke-machine","objectset.rio.cattle.io/owner-gvk":"rke.cattle.io/v1, Kind=RKEBootstrap","objectset.rio.cattle.io/owner-name":"custom-5402c6b1ed11","objectset.rio.cattle.io/owner-namespace":"fleet-local","rke.cattle.io/join-url":""}{"objectset.rio.cattle.io/applied":"H4sIAAAAAAAA/4xTzW7bPBB8lQ97Fv1JIhWRAnop2lMOBYKil7qH5WodM6ZIg6STFobfvZATAc2P0x7FnRnOjoZHmLjgiAVhOAKGEAsWF0OeP6O9YyqZyyq5uCIsxfPKxf/dCAOkHYsJaesCQ3URGh8CJ3F7v3tk/DG5b6r/rl0YP9xcf/4YY8kl4f6vSgEnhgHokEuchB0lat12bdPLf6LmPdLM33jmInwk9FC9MHYXXRCH5GGAVzOPln2GAY5r2GK651w4bcnNswkD3vK4hmENJR14DSc4VUCJz4l+dRPngtMehnDwvoJF6wjkD7PO6qfY6TxLPR0s2y4+L9wIw/m+dwLYYt7CALK+qlstpZWtbSRZVat2Y60xShKpXtuarDRd82rtC36eg57KIM7tqIlZK6OJuk52zHWvFanGmk53tSXJ2m66tm163SPa5soYMsb2hmRXj526KP7u/39OeYhpx0mk6HlJ6FTBZYGlzWLvMcAj8u2+nPt0wxtOHIgzDN+PgHv3jVN2MbzRdKjA+ki7LzPxE3suZ9zsqQKKoaToPaflZOfCnOGLh/Hu6odz6qPeWIU9iqsRWai+H4WWPQoaDdvW2I1SCKcfpwrKrz2/MvosgNPvAAAA///dmhgPGwQAAA","objectset.rio.cattle.io/id":"rke-machine","objectset.rio.cattle.io/owner-gvk":"rke.cattle.io/v1, Kind=RKEBootstrap","objectset.rio.cattle.io/owner-name":"custom-bd3a88252173","objectset.rio.cattle.io/owner-namespace":"fleet-local","rke.cattle.io/join-url":"","rke.cattle.io/labels":"{\"harvesterhci.io/managed\":\"true\"}"}

# "rke.cattle.io/join-url":"" 
```

The machines are also in the `Provisioning` phase:
```
harvester-node-0:/var/log # kubectl get machines.cluster.x-k8s.io -A
NAMESPACE     NAME                  CLUSTER   NODENAME   PROVIDERID   PHASE          AGE     VERSION
fleet-local   custom-5402c6b1ed11   local                             Provisioning   3h20m
fleet-local   custom-bd3a88252173   local                             Provisioning   3h10m
```

Looks like Node reference can't be established.
```
harvester-node-0:/var/log # kubectl describe machines.cluster.x-k8s.io custom-bd3a88252173 -n fleet-local
...
Status:
  Bootstrap Ready:  true
  Conditions:
    Last Transition Time:  2022-03-10T05:37:31Z
    Message:               1 of 2 completed
    Reason:                WaitingForInfrastructure
    Severity:              Info
    Status:                False
    Type:                  Ready
    Last Transition Time:  2022-03-10T05:37:31Z
    Status:                True
    Type:                  BootstrapReady
    Last Transition Time:  2022-03-10T05:37:29Z
    Reason:                WaitingForInfrastructure
    Severity:              Info
    Status:                False
    Type:                  InfrastructureReady
    Last Transition Time:  2022-03-10T05:37:29Z
    Reason:                WaitingForNodeRef
    Severity:              Info
    Status:                False
    Type:                  NodeHealthy
  Last Updated:            2022-03-10T05:37:31Z
  Observed Generation:     2
  Phase:                   Provisioning

```

**Comment by Oats87** [comment]
@guangbochen What is the health of the cluster? Are you able to `kubectl get pod -A` in it to figure out if everything came up correctly?

If the `cattle-cluster-agent` does not come up cleanly, the CAPI controllers + v2prov controllers will be unable to continue paving the cluster into place.

**Comment by thedadams** [comment]
The issue here seems to be that system-agent can't communicate with the Rancher server. This is the last log in the file for system-agent:
```
Mar 10 05:29:54 harvester-node-0 rancher-system-agent[11709]: E0310 05:29:54.771936   11709 reflector.go:139] pkg/mod/github.com/rancher/client-go@v0.21.0-rancher.1/tools/cache/reflector.go:168: Failed to watch *v1.Secret: failed to list *v1.Secret: Get "https://10.53.27.204/api/v1/namespaces/fleet-local/secrets?fieldSelector=metadata.name%3Dcustom-5402c6b1ed11-machine-plan&resourceVersion=5437": dial tcp 10.53.27.204:443: connect: connection refused
```
I am not entirely sure why that would happen. If I understand correctly, system-agent is running on the same machine as the Rancher management cluster. I know that `rancherd` sets the Rancher cluster to use host port, so maybe that is causing an issue?

I am honestly unsure what has changed recently that would cause this problem.

**Comment by bk201** [comment]
> @guangbochen What is the health of the cluster? Are you able to `kubectl get pod -A` in it to figure out if everything came up correctly?

@Oats87 The pods look good. 
[get_all.txt](https://github.com/harvester/harvester/files/8228489/get_all.txt)


> If the `cattle-cluster-agent` does not come up cleanly, the CAPI controllers + v2prov controllers will be unable to continue paving the cluster into place.

Do you mean rancher-system-agent? it's stuck with a connection error:
[rancher-system-agent.log](https://github.com/harvester/harvester/files/8228514/rancher-system-agent.log)

```
Mar 10 05:29:54 harvester-node-0 rancher-system-agent[11709]: E0310 05:29:54.771936   11709 reflector.go:139] pkg/mod/github.com/rancher/client-go@v0.21.0-rancher.1/tools/cache/reflector.go:168: Failed to watch *v1.Secret: failed to list *v1.Secret: Get "https://10.53.27.204/api/v1/namespaces/fleet-local/secrets?fieldSelector=metadata.name%3Dcustom-5402c6b1ed11-machine-plan&resourceVersion=5437": dial tcp 10.53.27.204:443: connect: connection refused
```

**Comment by bk201** [comment]
Just want to add this doesn't always happen. In most of my deployments, the joining works.
But it's more easily to happen in our CI test.

**Comment by bk201** [comment]
@Oats87 @thedadams We find the CAPI machine is stuck at `WaitingForNodeRef` condition and it might be because the `Spec.ProviderID` in `CustomMachine` objects are empty. By digging rancher's code, we can't see who sets the ProviderID, any idea what component might set it? Thanks!

**Comment by thedadams** [comment]
@bk201 The changes have been merged into Rancher.

**Comment by bk201** [comment]
Thanks @thedadams, we'll bump the version and test it.

**Comment by guangbochen** [comment]
This is a regression bug introduced after bumping the embedded Rancher from `v2.6.3` to the `v2.6-head` and it is fixed via https://github.com/harvester/harvester-installer/pull/246 and can be verified using the `v1.0.1-rc1` release.

**To Test:**
- other nodes should be able to join and form a cluster via the iPXE installation.
-  other nodes should be able to join and form a cluster via ISO installation.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.0.1-rc`. Close this issue.

### Result
1. Other nodes can join and form a 3 nodes cluster via ISO installation.

    ```
    harvester-node01-101rc1:~ # kubectl get nodes
    NAME STATUS ROLES AGE VERSION
    harvester-node01-101rc1 Ready control-plane,etcd,master 57m v1.22.7+rke2r1
    harvester-node02-101rc1 Ready control-plane,etcd,master 26m v1.22.7+rke2r1
    harvester-node03-101rc1 Ready control-plane,etcd,master 10m v1.22.7+rke2r1
    ```
    ```
    harvester-node01-101rc1:~ # kubectl get machines.cluster.x-k8s.io -A
    NAMESPACE NAME CLUSTER NODENAME PROVIDERID PHASE AGE VERSION
    fleet-local custom-664d2fa99436 local harvester-node03-101rc1 rke2://harvester-node03-101rc1 Running 13m
    fleet-local custom-b81eb7950700 local harvester-node02-101rc1 rke2://harvester-node02-101rc1 Running 28m
    fleet-local custom-c38430434754 local harvester-node01-101rc1 rke2://harvester-node01-101rc1 Running 58m
    ```

2. Other nodes can join and form a 4 nodes cluster vial iPXE installation.
    ```
    harvester-node-1:~ # kubectl get nodes
    NAME STATUS ROLES AGE VERSION
    harvester-node-0 Ready control-plane,etcd,master 15h v1.22.7+rke2r1
    harvester-node-1 Ready control-plane,etcd,master 15h v1.22.7+rke2r1
    harvester-node-2 Ready control-plane,etcd,master 15h v1.22.7+rke2r1
    harvester-node-3 Ready <none> 14h v1.22.7+rke2r1
    ```

    ```
    harvester-node-1:~ # kubectl get machines.cluster.x-k8s.io -A
    NAMESPACE NAME CLUSTER NODENAME PROVIDERID PHASE AGE VERSION
    fleet-local custom-06ff196da495 local harvester-node-0 rke2://harvester-node-0 Running 15h
    fleet-local custom-8818d6f1291f local harvester-node-2 rke2://harvester-node-2 Running 15h
    fleet-local custom-901374b2ac2a local harvester-node-1 rke2://harvester-node-1 Running 15h
    fleet-local custom-ae3c7b83fbfc local harvester-node-3 rke2://harvester-node-3 Running 14h
    ```

### Test Information
* Test Environment: 3 ~ 4 nodes harvester on provo bare machine
* Harvester version: v1.0.1-rc1

### Verify Steps
1. ISO install a 3 nodes harvester cluster
2. Check harvester can install correctly
3. ssh to node 1 machine
4. Check can display node by running command
  - `kubectl get nodes`
  - `kubectl get machines.cluster.x-k8s.io -A`
5. Use iPXE vagrant example to install a 4 nodes harvester cluster
6. Check harvester can install correctly
7. ssh to node 1 machine
8. Check can display node by running command
  - `kubectl get nodes`
  - `kubectl get machines.cluster.x-k8s.io -A`
